tiledb/api/src/range.rs

```diff
@@ -0,0 +1,841 @@
+    pub fn check_datatype(&self, datatype: Datatype) -> TileDBResult<()> {
+        check_datatype!(self, datatype);
+        Ok(())
+    }
+}
+
+macro_rules! var_value_range_from {
+    ($($V:ident : $U:ty),+) => {
+        $(
+            impl From<(Box<[$U]>, Box<[$U]>)> for VarValueRange {
+                fn from(value: (Box<[$U]>, Box<[$U]>)) -> VarValueRange {
+                    VarValueRange::$V(value.0, value.1)
+                }
+            }
+        )+
+    }
+}
+
+var_value_range_from!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
+var_value_range_from!(Int8: i8, Int16: i16, Int32: i32, Int64: i64);
+var_value_range_from!(Float32: f32, Float64: f64);
+
+impl From<(&str, &str)> for VarValueRange {
+    fn from(value: (&str, &str)) -> VarValueRange {
+        let start = value.0.as_bytes().to_vec().into_boxed_slice();
+        let end = value.1.as_bytes().to_vec().into_boxed_slice();
+        VarValueRange::UInt8(start, end)
+    }
+}
+
+impl From<&[&str; 2]> for VarValueRange {
+    fn from(value: &[&str; 2]) -> VarValueRange {
+        let start = value[0].as_bytes().to_vec().into_boxed_slice();
+        let end = value[1].as_bytes().to_vec().into_boxed_slice();
+        VarValueRange::UInt8(start, end)
+    }
+}
+
+#[macro_export]
+macro_rules! var_value_range_go {
+    ($expr:expr, $DT:ident, $start:pat, $end:pat, $then:expr) => {
+        match $expr {
+            VarValueRange::UInt8($start, $end) => {
+                type $DT = u8;
+                $then
+            }
+            VarValueRange::UInt16($start, $end) => {
+                type $DT = u16;
+                $then
+            }
+            VarValueRange::UInt32($start, $end) => {
+                type $DT = u32;
+                $then
+            }
+            VarValueRange::UInt64($start, $end) => {
+                type $DT = u64;
+                $then
+            }
+            VarValueRange::Int8($start, $end) => {
+                type $DT = i8;
+                $then
+            }
+            VarValueRange::Int16($start, $end) => {
+                type $DT = i16;
+                $then
+            }
+            VarValueRange::Int32($start, $end) => {
+                type $DT = i32;
+                $then
+            }
+            VarValueRange::Int64($start, $end) => {
+                type $DT = i64;
+                $then
+            }
+            VarValueRange::Float32($start, $end) => {
+                type $DT = f32;
+                $then
+            }
+            VarValueRange::Float64($start, $end) => {
+                type $DT = f64;
+                $then
+            }
+        }
+    };
+}
+
+#[derive(Clone, Deserialize, Serialize, PartialEq)]
+pub enum Range {
+    Single(SingleValueRange),
+    Multi(MultiValueRange),
+    Var(VarValueRange),
+}
+
+impl Range {
+    // N.B. This is not `check_field_compatibility` because dimensions have
+    // restrictions on their cell_val_num that don't apply to attributes.
```

Comment:
We also don't apply subarray ranges to attributes... right?

Reply:
Not at all. Though I tend to start implementing from the point of view of general principles. I.e.,, "I already have to write X to do Y, how hard is it to make it also do Z, Q, R". I figured, well, attributes have all the things for these checks and we have the field API so why not just `check_field_compatibility` at first. Then after I had it mostly written and was checking core for extra limitations I might be missing I realized that wasn't as general so I left a note.